### DISCUSSION TITLE
Easily switch between .env files

### DISCUSSION THREAD

**Discussion by saibotsivad** [usebruno/bruno#828] [Ideas]
## The Praise

The approach to [managing secrets](https://docs.usebruno.com/secrets-management/dotenv-file.html) in a `.env` is really great, it's been pretty cool to be able to keep an `.env.example` file in the git repo.

Also, from within Bruno, the ability to switch between environments is great. You just pick one from the dropdown selector!

(I also suggest adding `.env*` and then `!.env.example` to the `.gitignore` so all `.env` files are ignored _except_ the example one.)

## The Pain

When storing secrets in the `.env` file, switching between environments actually requires the following steps (for example switching from staging to production):

- From within Bruno, pick "Production" from the dropdown selector
- Go to the folder with the `.env` files
- Rename `.env` to `.env.staging`
- Rename `.env.production` to `.env`

That's a lot of work, and annoying!

## Proposal?

There could be a way from in Bruno to pick which `.env*` file to use? That adds another dropdown selector, which isn't a great UX but could be okay.

Or perhaps a way to associate a Bruno environment with a specific `.env` file? For example in a file `environments/Production.bru` there could be:

```bru
env {
  file: .env.production
}
vars {
  apiKey: {{process.env.API_KEY}}
}
```

This doesn't feel like a great solution either, because a user might desire to point to a different file than `.env.production` and then you'd have to remember to never check in the `env { file }` part of the `Production.bru` file.

Are there other ways to manage this?

**Comment by helloanoop** [comment]
Hey @saibotsivad !

One approach is utilising prefixes (e.g., DEV, QA, STG) to distinguish between different environments.

`.env`
```env
DEV_AWS_ACCESS_KEY = secret
QA_AWS_ACCESS_KEY = secret
STG_AWS_ACCESS_KEY = secret
```

And then in your envs
`environments/dev.bru`
```
vars {
  AWS_ACCESS_KEY: {{process.env.DEV_AWS_ACCESS_KEY}}
}
```
`environments/qa.bru`
```
vars {
  AWS_ACCESS_KEY: {{process.env.AWS_ACCESS_KEY}}
}
```

In this approach, your app would use a single variable `AWS_ACCESS_KEY`, but these would be resolved to different values depending on the environment selected in the dropdown. 

*Note*: While each environment may contain numerous variables, only a small subset of them actually require secrecy.
So, you could save the regular variables inside the usual environment configuration, and only the secret ones in the `.env` file.

**Reply by saibotsivad** [reply]
Yeah, I end up doing this, but it's clunky and feels like there could be a more elegant approach.

Having multiple `.env` files is a common problem, so this in itself isn't specific to Bruno, but I think the pain unique to Bruno is that changing environments becomes a multi-step process. Most projects I work in that use `.env` files make use of [env-cmd](https://www.npmjs.com/package/env-cmd), which lets you simply pass in a flag to point to a different `.env` file, so switching environments is a single flag.

One other solution I was thinking about is if each Bruno environment was *optionally* associated with an *additional* named `.env` but that association would not be a setting stored in the folder.

So suppose you have some normal `environment/PROD.bru` and `environment/STAGE.bru` files, and then you'd have some a selector inside the Environment modal, like this:

![bruno-env-option](https://github.com/usebruno/bruno/assets/91254/c53a0727-1f08-45c6-8842-5f26b83ab1c3)

**Comment by xalvarez** [comment]
I feel the same pain and like the ideas exposed here by @saibotsivad, I just wanted to add one more, which would be to optionally read an .env file found in an environment folder. This could take precedence over an .env file created at the collection root.

**Reply by hellodave76** [reply]
This is what I just tried, thinking surely this will just work, right? Unfortunately not... yet!
+1 for some kind of native support for handling multiple .env files 🙏

**Comment by yluom** [comment]
Hello, any updates about this topic ? I think it is a nice improvement that will help people migrate to bruno !

**Comment by leaumar** [comment]
My team and I only recently started using bruno and only today found out about these handlebars-style env vars. But our first and immediate grievances were indeed
- why can't we make `.env.acc`, `.env.prod` etc instead of a single `.env` that complicates variable naming by having to prefix the env in every value
- and why can't bruno just autodiscover the envs from our `.env.*` files and let us select them, and fill the values in the requests directly from the file without the additional mapping layer of `jwt.secret=process.env.JWT_SECRET`

On a sidenote, using a .env file doesn't even work for us right now. We're following the docs and our experience with .env files, and bruno even resolves the secrets when we hover over the handlebar names in requests, but the actual request is fired without replacing the `{{foo}}` by their values. Is there some requirement to using this feature that we're missing?

**Reply by mrwensveen** [reply]
As to your second point: you don't have to create environments and map variables. You can just use environment variables directly (loaded from .env as well as the actual runtime enviroment) using the handlebar syntax. The mapping step is useful when you want to map bruno variables to different enviroment variables depending on the bruno environment (e.g. `jwt_secret={{process.env.PRD_JWT_SECRET}}`)

HTH

**Comment by mj-h** [comment]
I gave up on `.env` files and just use environments, because I have to switch between environments a lot.

I don't check my environments into the repo (my `.gitignore` line is `**/environments/*.bru`), but I do maintain a `environments/sample-env.bru.sample` file so that my colleagues know what environment variables they have to define. For complex scenarios, I had Claude/ChatGPT write bash helper scripts for me that autogenerate environments for me.

Other colleagues of mine commit their environments to the repo, but they use the "is secret" checkbox, so that the actual value is not saved in the environment. It's a super convenient feature, but it does not work for me, because I need those values in the cli runner as well, and currently, the cli runner has no way to read the secrets from the UI-secrets-store.

![image](https://github.com/user-attachments/assets/a204deac-ebab-4f55-9837-7f9fce1de8fc)

**Comment by medQ-Tommy** [comment]
Add me to the list of those who thinks the .env is a great idea that isn't well implemented yet. There really needs to be support for multiple .env files, not creating one giant env with a ton of different versions of the the variables in it.

**Comment by matthiasblomme** [comment]
How about just supporting a .env file per environment and a global one? So that these can be combined and overwritten?

**Comment by filoscoder** [comment]
+1 for `.env.*` support 👍🏼

**Comment by chadbirch** [comment]
The way [ASP.NET Core manages application settings](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-9.0#default-application-configuration-sources), specifically around the `appsettings` files, would be a nice approach here. 

They have an `appsettings.json` that gets processed first. then they have an `appsettings.{ENVIRONMENT_NAME}.json` that gets processed next. It overrides values defined in the `appsettings.json` and adds the remaining values as new values.

In this case, process the `.env` file. Then, if you're environment is set to QA_US, load the `.env.QA_US` if it exists. You'll need to define some logic for spaces and special characters since Environment Names may contain characters not allowed in a filename for the OS, but I believe this approach would be solve the issue that the OP brought up.

**Comment by Adam-Coetzee** [comment]
A little sad to see this is not a feature in Bruno yet. Loading .env-<bruno-env> depending on selected bruno env and dynamically reading  the correct corresponding env file.

**Reply by Adam-Coetzee** [reply]
I feel this to be a must for convenience's sake, the value will be making it easier to sell for normal dev workflows! How can I convince my colleagues with this baffling omission

**Comment by anusree-bruno** [comment]
To add an update on this — there are a couple of PRs currently attempting to address this functionality: [#5004](https://github.com/usebruno/bruno/pull/5004) and [#2578](https://github.com/usebruno/bruno/pull/2578).
This is also being tracked in the related issue: [#409](https://github.com/usebruno/bruno/issues/409).

**Comment by sklakegg** [comment]
This would be great. Tools like dotenvx also work better when allowing postfixes like .env.dev and .env.prod
